api/envoy/service/mutation/v3alpha/mutation.proto

```diff
@@ -0,0 +1,302 @@
+import "envoy/service/auth/v3/attribute_context.proto";
+import "envoy/type/v3/http_status.proto";
+
+import "google/protobuf/duration.proto";
+import "google/protobuf/struct.proto";
+
+import "udpa/annotations/status.proto";
+import "validate/validate.proto";
+
+option java_package = "io.envoyproxy.envoy.service.mutation.v3alpha";
+option java_outer_classname = "MutationProto";
+option java_multiple_files = true;
+option java_generic_services = true;
+option (udpa.annotations.file_status).work_in_progress = true;
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: External Mutation Service]
+
+// A service that can mutate HTTP requests and responses.
+// The overall mutation protocol works like this:
+//
+// 1. Envoy sends to the service information about the HTTP request.
+// 2. The service sends back a MutationResponse message that directs Envoy
+//    to either stop processing, continue without it, or send it the
+//    next chunk of the message body.
+// 3. If so requested, Envoy sends the server chunks of the message body,
+//    or the entire body at once. In either case, the server sends back
+//    a MutationResponse after each message it receives.
+// 4. If so requested, Envoy sends the server the HTTP trailers,
+//    and the server sends back a MutationResponse.
+// 5. At this point, request processing is done, and we pick up again
+//    at step 1 when Envoy receives a response from the upstream server.
+// 6. At any point above, if the server closes the gRPC stream cleanly,
+//    then Envoy proceeds without consulting the server.
+// 7. At any point above, if the server closes the gRPC stream with an error,
+//    then Envoy returns a 500 error to the client, unless the filter
+//    was configured to ignore errors.
+//
+// In other words, the process is a request/response conversation, but
+// using a gRPC stream to make it easier for the client and server to
+// maintain state. The exception is the "SEND_BODY_ASYNC" command,
+// which requests that the server keep sending body chunks without waiting
+// for a response.
+
+service Mutation {
+  // This begins the bidirectional stream that Envoy will use to
+  // give the server control over what the filter does. The actual
+  // protocol is described by the MutationRequest and MutationResponse
+  // messages below.
+  rpc Mutate(stream MutationRequest) returns (stream MutationResponse) {
+  }
+}
+
+// This represents the different types of messages that Envoy can send
+// to a mutation server.
+// [#next-free-field: 8]
+message MutationRequest {
+  // If true, then Envoy requires that the server send one MutationResponse message
+  // back up the stream in response to this message -- otherwise, Envoy will
+  // never resume processing the filter chain. In practice, this will always
+  // be "true" unless the server or the user has configured the filter
+  // to run in "async mode."
+  bool response_required = 1;
```

Comment:
Would it better to just send the mode to the server? I think the intention would be more clear?

Reply:
I'm trying to avoid making every server author understand a complex set of calculations. The idea is that the proxy sends one message to the server, and there's a flag right there that tells the server either that the message was informational, or that the proxy needs to respond in a timely way (because the server has presumably suspended filter execution while it waits for a response).